# checked-requires-onchange-or-readonly

> **Keywords:** React, controlled input, checkbox, radio, onChange, readOnly, form handling, ESLint rule, LLM-optimized

Ensures controlled inputs with `checked` or `value` props have an `onChange` handler or `readOnly` attribute. This rule is part of [`@forge-js/eslint-plugin-llm-optimized`](https://www.npmjs.com/package/@forge-js/eslint-plugin-llm-optimized) and provides LLM-optimized error messages.

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (React best practice)                                        |
| **Auto-Fix**   | âŒ No (requires handler implementation)                              |
| **Category**   | React                                                                |
| **ESLint MCP** | âœ… Optimized for ESLint MCP integration                              |
| **Best For**   | React form handling, controlled components                           |

## Rule Details

```mermaid
%%{init: {
  'theme': 'base',
  'themeVariables': {
    'primaryColor': '#f8fafc',
    'primaryTextColor': '#1e293b',
    'primaryBorderColor': '#334155',
    'lineColor': '#475569'
  }
}}%%
flowchart TD
    A[ğŸ” Check input/textarea/select] --> B{Has checked or value?}
    B -->|âŒ No| C[âœ… Pass]
    B -->|âœ… Yes| D{Has onChange?}
    D -->|âœ… Yes| C
    D -->|âŒ No| E{Has readOnly?}
    E -->|âœ… Yes| C
    E -->|âŒ No| F[âš ï¸ Report: Missing handler]
    
    classDef startNode fill:#f0fdf4,stroke:#16a34a,stroke-width:2px,color:#1f2937
    classDef errorNode fill:#fef2f2,stroke:#dc2626,stroke-width:2px,color:#1f2937
    classDef processNode fill:#eff6ff,stroke:#2563eb,stroke-width:2px,color:#1f2937
    
    class A startNode
    class F errorNode
    class C processNode
```

### Why This Matters

| Issue                   | Impact                          | Solution                       |
| ----------------------- | ------------------------------- | ------------------------------ |
| ğŸš« **Uncontrolled**     | User can't interact with input  | Add onChange handler           |
| âš ï¸ **React Warning**    | Console warnings in development | Properly handle controlled state |
| ğŸ› **Silent Failures**  | Form doesn't work as expected   | Use readOnly for display-only  |
| ğŸ“‹ **State Sync**       | State and UI out of sync        | Implement proper handlers      |

## Examples

### âŒ Incorrect

```jsx
// Missing onChange - user can't toggle checkbox
<input type="checkbox" checked={isChecked} />

// Missing onChange - user can't type
<input type="text" value={name} />

// Missing handler on textarea
<textarea value={description} />

// Select without handler
<select value={selectedOption}>
  <option value="a">A</option>
</select>
```

### âœ… Correct

```jsx
// With onChange handler
<input 
  type="checkbox" 
  checked={isChecked} 
  onChange={(e) => setIsChecked(e.target.checked)} 
/>

// With onChange for text input
<input 
  type="text" 
  value={name} 
  onChange={(e) => setName(e.target.value)} 
/>

// With readOnly for display-only
<input type="text" value={displayValue} readOnly />

// Properly controlled textarea
<textarea 
  value={description} 
  onChange={(e) => setDescription(e.target.value)} 
/>

// Using onInput as alternative
<input 
  type="text" 
  value={search} 
  onInput={(e) => setSearch(e.target.value)} 
/>
```

## Configuration

```javascript
{
  rules: {
    '@forge-js/checked-requires-onchange-or-readonly': 'error'
  }
}
```

## When to Use readOnly

| Scenario                       | Use `readOnly` | Use `onChange`  |
| ------------------------------ | -------------- | --------------- |
| Display calculated values      | âœ…             | âŒ              |
| Show server-provided data      | âœ…             | âŒ              |
| User-editable forms            | âŒ             | âœ…              |
| Toggle checkboxes              | âŒ             | âœ…              |

## Related Rules

- [`jsx-no-bind`](./jsx-no-bind.md) - Prevent binding in JSX props
- [`no-direct-mutation-state`](./no-direct-mutation-state.md) - Prevent direct state mutation

## Further Reading

- **[Controlled Components](https://react.dev/reference/react-dom/components/input#controlling-an-input-with-a-state-variable)** - React documentation
- **[Forms in React](https://react.dev/learn/responding-to-events)** - React forms guide
- **[ESLint MCP Setup](https://eslint.org/docs/latest/use/mcp)** - Enable AI assistant integration
