# jsx-key

> **Keywords:** React, JSX, key prop, reconciliation, lists, map, iteration, ESLint rule, performance, LLM-optimized

Detect missing or problematic React keys that could break reconciliation. This rule is part of [`@forge-js/eslint-plugin-llm-optimized`](https://www.npmjs.com/package/@forge-js/eslint-plugin-llm-optimized) and provides LLM-optimized error messages with suggestions.

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Error (correctness)                                                  |
| **Auto-Fix**   | ğŸ’¡ Suggests fixes                                                    |
| **Category**   | React                                                                |
| **ESLint MCP** | âœ… Optimized for ESLint MCP integration                              |
| **Best For**   | All React/JSX projects                                               |

## Rule Details

```mermaid
%%{init: {
  'theme': 'base',
  'themeVariables': {
    'primaryColor': '#f8fafc',
    'primaryTextColor': '#1e293b',
    'primaryBorderColor': '#334155',
    'lineColor': '#475569',
    'c0': '#f8fafc',
    'c1': '#f1f5f9',
    'c2': '#e2e8f0',
    'c3': '#cbd5e1'
  }
}}%%
flowchart TD
    A[ğŸ” JSX in Iterator] --> B{Has key prop?}
    B -->|âŒ No| C[âŒ Report: Missing Key]
    B -->|âœ… Yes| D{Key Quality?}
    
    D -->|ğŸ”¢ Index| E[âš ï¸ Warn: Unstable Key]
    D -->|âœ… Stable ID| F[âœ… Pass]
    D -->|ğŸ” Duplicate| G[âŒ Report: Duplicate Key]
    
    classDef startNode fill:#f0fdf4,stroke:#16a34a,stroke-width:2px,color:#1f2937
    classDef errorNode fill:#fef2f2,stroke:#dc2626,stroke-width:2px,color:#1f2937
    classDef warnNode fill:#fef3c7,stroke:#f59e0b,stroke-width:2px,color:#1f2937
    classDef processNode fill:#eff6ff,stroke:#2563eb,stroke-width:2px,color:#1f2937
    
    class A startNode
    class C,G errorNode
    class E warnNode
    class F processNode
```

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| ğŸ”„ **Reconciliation**     | React can't track elements      | Add unique keys           |
| âš¡ **Performance**        | Unnecessary re-renders          | Stable keys               |
| ğŸ› **State Bugs**         | Wrong component gets state      | Use item IDs, not indexes |
| ğŸ¨ **Animation Issues**   | Elements animate incorrectly    | Consistent key identity   |

## Configuration

| Option             | Type      | Default | Description                              |
| ------------------ | --------- | ------- | ---------------------------------------- |
| `warnUnstableKeys` | `boolean` | `true`  | Warn about potentially unstable keys     |

## Examples

### âŒ Incorrect

```tsx
// Missing key
function UserList({ users }) {
  return (
    <ul>
      {users.map(user => (
        <li>{user.name}</li>  // âŒ Missing key
      ))}
    </ul>
  );
}

// Using index as key (unstable)
function ItemList({ items }) {
  return (
    <ul>
      {items.map((item, index) => (
        <li key={index}>{item.name}</li>  // âš ï¸ Unstable key
      ))}
    </ul>
  );
}
```

### âœ… Correct

```tsx
// Using unique ID as key
function UserList({ users }) {
  return (
    <ul>
      {users.map(user => (
        <li key={user.id}>{user.name}</li>  // âœ… Stable unique key
      ))}
    </ul>
  );
}

// Using compound key when no ID available
function ItemList({ items }) {
  return (
    <ul>
      {items.map(item => (
        <li key={`${item.category}-${item.name}`}>
          {item.name}
        </li>  // âœ… Unique compound key
      ))}
    </ul>
  );
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    '@forge-js/jsx-key': 'error'
  }
}
```

### Disable Unstable Key Warnings

```javascript
{
  rules: {
    '@forge-js/jsx-key': ['error', {
      warnUnstableKeys: false
    }]
  }
}
```

## Key Best Practices

### Choosing Keys

| Source              | Quality    | Example                              |
| ------------------- | ---------- | ------------------------------------ |
| Database ID         | âœ… Best    | `key={user.id}`                      |
| Unique field        | âœ… Good    | `key={item.slug}`                    |
| Compound unique     | âœ… Good    | `key={\`${cat}-${name}\`}`           |
| UUID/nanoid         | âš ï¸ OK      | `key={generateId()}`                 |
| Array index         | âŒ Avoid   | `key={index}`                        |
| Random number       | âŒ Never   | `key={Math.random()}`                |

### When Index Keys Are Acceptable

```tsx
// âœ… Static list that never reorders/filters
const WEEKDAYS = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri'];
<ul>
  {WEEKDAYS.map((day, i) => <li key={i}>{day}</li>)}
</ul>

// âŒ Dynamic list - DON'T use index
{users.map((user, i) => <UserCard key={i} user={user} />)}  // Bad!
{users.map(user => <UserCard key={user.id} user={user} />)} // Good!
```

## Common Patterns

### Fragment Keys

```tsx
// When using fragments in lists, key goes on Fragment
function DataList({ data }) {
  return (
    <>
      {data.map(item => (
        <React.Fragment key={item.id}>
          <dt>{item.term}</dt>
          <dd>{item.definition}</dd>
        </React.Fragment>
      ))}
    </>
  );
}
```

### Nested Lists

```tsx
function NestedList({ categories }) {
  return categories.map(category => (
    <div key={category.id}>
      <h2>{category.name}</h2>
      <ul>
        {category.items.map(item => (
          <li key={item.id}>{item.name}</li>  // Separate key namespace
        ))}
      </ul>
    </div>
  ));
}
```

## When Not To Use

| Scenario                    | Recommendation                              |
| --------------------------- | ------------------------------------------- |
| ğŸ§ª **Testing/Prototypes**   | Consider allowing index keys temporarily    |
| ğŸ“Š **Static content**       | Index keys may be acceptable                |
| ğŸ”„ **No reordering**        | Index keys work for append-only lists       |

## Comparison with Alternatives

| Feature              | jsx-key             | eslint-plugin-react | jsx-ally           |
| -------------------- | ------------------- | ------------------- | ------------------ |
| **Missing key**      | âœ… Yes              | âœ… Yes              | âŒ No              |
| **Unstable keys**    | âœ… Configurable     | âš ï¸ Limited          | âŒ No              |
| **LLM-Optimized**    | âœ… Yes              | âŒ No               | âŒ No              |
| **ESLint MCP**       | âœ… Optimized        | âŒ No               | âŒ No              |
| **Suggestions**      | âœ… Yes              | âš ï¸ Limited          | âŒ No              |

## Related Rules

- [`react-no-inline-functions`](./react-no-inline-functions.md) - Performance optimization
- [`react-render-optimization`](./react-render-optimization.md) - Render performance

## Further Reading

- **[React Keys Documentation](https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key)** - Official React docs
- **[Why React Keys Matter](https://kentcdodds.com/blog/understanding-reacts-key-prop)** - Kent C. Dodds article
- **[Reconciliation](https://react.dev/learn/preserving-and-resetting-state)** - How React updates the DOM
- **[ESLint MCP Setup](https://eslint.org/docs/latest/use/mcp)** - Enable AI assistant integration

