# react-render-optimization

> **Keywords:** react render optimization, performance, ESLint rule, JavaScript, TypeScript

ESLint Rule: react-render-optimization with LLM-optimized suggestions and auto-fix capabilities.

## Quick Summary

| Aspect         | Details                                      |
| -------------- | -------------------------------------------- |
| **Severity**   | Error (code quality)                        |
| **Auto-Fix**   | ‚ùå No                                        |
| **Category**   | Performance |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration      |
| **Best For**   | Production applications                      |
| **Suggestions** | ‚úÖ 6 available           |

## Rule Details

```mermaid
%%{init: {
  'theme': 'base',
  'themeVariables': {
    'primaryColor': '#f8fafc',
    'primaryTextColor': '#1e293b',
    'primaryBorderColor': '#334155',
    'lineColor': '#475569',
    'c0': '#f8fafc',
    'c1': '#f1f5f9',
    'c2': '#e2e8f0',
    'c3': '#cbd5e1'
  }
}}%%
flowchart TD
    A[üîç Detect react render optimization] --> B{Valid pattern?}
    B -->|‚ùå No| C[üö® Report violation]
    B -->|‚úÖ Yes| D[‚úÖ Pass]

    classDef startNode fill:#f0fdf4,stroke:#16a34a,stroke-width:2px,color:#1f2937
    classDef errorNode fill:#fef2f2,stroke:#dc2626,stroke-width:2px,color:#1f2937

    class A startNode
    class C errorNode
```

### Why This Matters

| Issue                | Impact                                | Solution                    |
| -------------------- | ------------------------------------- | --------------------------- |
| üîí **Security/Code Quality** | [Specific issue] | [Solution approach] |
| üêõ **Maintainability** | [Impact] | [Fix] |
| ‚ö° **Performance**   | [Impact] | [Optimization] |

## Configuration

**No configuration options available.**

## Examples

### ‚ùå Incorrect

```typescript
// Example of incorrect usage
```

### ‚úÖ Correct

```typescript
// Example of correct usage
```

## Configuration Examples

### Basic Usage

```javascript
// eslint.config.mjs
export default [
  {
    rules: {
      '@forge-js/react-render-optimization': 'error',
    },
  },
];
```

## LLM-Optimized Output

```
üö® react render optimization | Description | MEDIUM
   Fix: Suggestion | Reference
```

## Related Rules

- [`rule-name`](./rule-name.md) - Description

## Further Reading

- **[Reference](https://example.com)** - Description
