# react-no-inline-functions

Prevent inline functions in React renders with performance metrics.

**üí° Provides suggestions** | **üîß Automatically fixable**

## Rule Details

Detects inline functions in React JSX that cause unnecessary re-renders, providing performance impact analysis.

## Configuration

| Option                  | Type      | Default | Description                                       |
| ----------------------- | --------- | ------- | ------------------------------------------------- |
| `allowInEventHandlers`  | `boolean` | `false` | Allow inline functions in event handlers          |
| `minArraySize`          | `number`  | `10`    | Minimum array size to report inline functions in `.map()` |

## Examples

### ‚ùå Incorrect

```typescript
function TodoList({ todos }: Props) {
  return (
    <div>
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          onDelete={() => deleteTodo(todo.id)}  // ‚ùå Inline function
        />
      ))}
    </div>
  );
}
```

### ‚úÖ Correct

```typescript
function TodoList({ todos }: Props) {
  const handleDelete = useCallback((todoId: string) => {
    deleteTodo(todoId);
  }, []);
  
  return (
    <div>
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          onDelete={() => handleDelete(todo.id)}  // ‚úÖ Using useCallback
        />
      ))}
    </div>
  );
}
```

## Configuration Examples

```javascript
{
  rules: {
    '@forge-js/react-no-inline-functions': ['warn', {
      allowInEventHandlers: true,  // Allow simple event handlers
      minArraySize: 5              // Only warn for large lists
    }]
  }
}
```

## Performance Impact

| Array Size | Re-renders | Impact  |
| ---------- | ---------- | ------- |
| 1-10       | Low        | üü¢ Minor |
| 11-100     | Medium     | üü° Moderate |
| 100+       | High       | üî¥ Significant |

## Related Rules

- [`react-class-to-hooks`](./react-class-to-hooks.md) - Migration to hooks

