# autocomplete-valid

> **Keywords:** autocomplete-valid, accessibility, ESLint rule, WCAG, a11y, React accessibility

Require valid autocomplete attribute. This rule is part of [`eslint-plugin-react-a11y`](https://www.npmjs.com/package/eslint-plugin-react-a11y) and provides LLM-optimized error messages with fix suggestions.

## Quick Summary

| Aspect | Details |
|--------|---------|
| **WCAG Criterion** | 1.3.5 Identify Input Purpose |
| **Severity** | Error/Warning (accessibility) |
| **Auto-Fix** | üí° Suggestions available |
| **Category** | Accessibility |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration |

## Rule Details

This rule helps ensure WCAG 1.3.5 compliance by enforcing: Require valid autocomplete attribute

### Why This Matters

| Issue | Impact | Standard |
|-------|--------|----------|
| ‚ôø **Accessibility** | Screen reader and assistive technology users affected | WCAG 1.3.5 |
| ‚öñÔ∏è **Legal** | ADA/Section 508 compliance risk | Legal Requirement |
| üîç **SEO** | Search engines prefer accessible sites | Best Practice |

## Examples

### ‚ùå Incorrect

```tsx
// Violation of autocomplete-valid
// See rule source for specific examples
```

### ‚úÖ Correct

```tsx
// Compliant with autocomplete-valid
// See rule source for specific examples
```

## Configuration

```javascript
// eslint.config.js
{
  rules: {
    'react-a11y/autocomplete-valid': 'error'
  }
}
```

## WCAG 2.1 Compliance

This rule helps satisfy:
- **1.3.5 Identify Input Purpose**: Require valid autocomplete attribute

## Related Rules

- See [RULES.md](../RULES.md) for all accessibility rules

## Further Reading

- **[WCAG 1.3.5](https://www.w3.org/WAI/WCAG21/Understanding/)** - WCAG guidelines
- **[WebAIM](https://webaim.org/)** - Accessibility resources
- **[MDN Accessibility](https://developer.mozilla.org/en-US/docs/Web/Accessibility)** - MDN documentation

## Version

This rule is available in `eslint-plugin-react-a11y` v1.0.0+
